### PR TITLE
*: make parser package dependency as small as possible

### PR DESCRIPTION
**Pull request by tiancaiamao**
### What problem does this PR solve? 

Make parser package's dependency as small as possible.

This is nearly the final step before we [move parser to a separate repository](https://github.com/pingcap/tidb/issues/7923)

parser

All package in that picture will be moved to this new repo https://github.com/pingcap/parser

### What is changed and how it works?

* `types` is split into two, a smaller one put into `parser/types` which does not have much dependency and seldom changes.
* define `NewDecimal/NewHexLiteral/NewBitLiteral` function and `types/parser_driver` provide the implementation

### Check List 

Tests 

 - No code

@lamxTyler @zz-jason

### CODE REVIEW

File: parser/misc.go [hunk 1]
```diff
@@ -584,7 +583,7 @@ func (s *Scanner) isTokenIdentifier(lit string, offset int) int {
 		}
 	}
 
-	checkBtFuncToken, tokenStr := false, hack.String(data)
```
Comment (winoros) [inline parser/misc.go]: Will this influence performance?

Comment (tiancaiamao) [reply parser/misc.go]: Perhaps. 
When parser is moved into a separate repository, optimize its performance would be another task. I suggest that we can get the first step done.

File: parser/types/etc.go [hunk 2]
```diff
@@ -0,0 +1,258 @@
...
+	KindMysqlTime:     "datetime",
+	KindInterface:     "interface",
+	KindMinNotNull:    "min_not_null",
+	KindMaxValue:      "max_value",
+	KindRaw:           "raw",
+	KindMysqlJSON:     "json",
+}
+
+// TypeStr converts tp to a string.
+func TypeStr(tp byte) (r string) {
```
Comment (lamxTyler) [inline parser/types/etc.go]: Seems like it is not used?

Comment (tiancaiamao) [reply parser/types/etc.go]: This function is used in `field_type.go`
I've remove some redundant code.

**Comment by jackysp** [comment]
Should we merge https://github.com/pingcap/parser/pull/1 with this one?

**Comment by tiancaiamao** [comment]
@jackysp  https://github.com/pingcap/parser/pull/1 is actually this code, with the import path revised

After this PR is merged, I will
1. update the pingcap/parser repo, make sure it's latest
2. remove the tidb/parser directory from tidb, import pingcap/parser and update vendor
3. remove vendor, migrate to go module

**Pull request merged**